sem: report an error for calls to erroneous routines

zerbina opened the pull request:
## Summary

If overload resolution picked the symbol of a routine whose definition (header or body) contains an error, the call expression was passed on, resulting in AST containing `nkError` nodes to possibly `transf` and `mirgen` (in the case of macro evaluation), crashing the compiler.

After overload resolution, `semOverloadedCall` now produces an `nkError` if the callee's definition contains an error (identified by the `TSym.ast` field containing an `nkError` node).

## Details

- update the `TSym.ast` of the produced symbol with the error AST in `semProcAux`, `semTemplateDef`, and `generateInstance`
- short-circuit `trackProc` if the body has an error. This fixes some compiler crashes due to errors reaching logic that they shouldn't, and also fixes errors in routine bodies being reported twice

There still exists logic that doesn't resolve macros and templates through `semOverloadedCall` call. In order to help with finding these, `nkError` nodes reaching `transf` or template expansion are now treated as an internal error.

### Misc

- fix a field-access defect in `semCustomPragma` that happened when `semOverloadedCall` returns an error

---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

## Notes for Reviewers
* this change unblocks #590
* the short-circuiting in `trackProc` fixes an issue that blocks #594 

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

zerbina committed:
sem: report an error for calls to erroneous routines

Summary
=======

If overload resolution picked the symbol of a routine whose definition
(header or body) contains an error, the call expression was passed on,
resulting in AST containing `nkError` nodes to possibly `transf` and
`mirgen` (in the case of macro evaluation), crashing the compiler.

After overload resolution, `semOverloadedCall` now produces an
`nkError` if the callee's definition contains an error (identified by
the `TSym.ast` field containing an `nkError` node).

Details
=======

Both `semProcAux` and `semTemplateDef` are adjusted to also update the
`TSym.ast` with the `nkError` node, otherwise the detection of a routine
with an error is not possible.

There still exists logic that doesn't resolve macros and templates
through `semOverloadedCall` call. In order to help in finding these,
`nkError` nodes reaching `transf` or template expansion are now treated
as an internal error.

Misc
----

- fix a field-access defect in `semCustomPragma` that happened when
  `semOverloadedCall` returns an error

saem commented on compiler/sem/seminst.nim:
```diff
@@ -393,6 +393,13 @@ proc generateInstance(c: PContext, fn: PSym, pt: TIdTable,
     if c.inGenericContext == 0:
       instantiateBody(c, n, fn.typ.n, result, fn)
+      if result.ast[bodyPos].kind == nkError:
+        # XXX: we also need to report the error here for now. Without the
+        #      ``localReport``, the error would never get reported
```
Yup, reworking the callsites to remove this is better as a separate PR.

saem commented on compiler/sem/transf.nim:
```diff
@@ -907,10 +906,7 @@ proc transform(c: PTransf, n: PNode): PNode =
-    result = n
-    c.graph.config.localReport(n)
-    return
+    unreachable("errors can't reach here")
```
I really like the idea of using `transf` as a border to force better layering.

zerbina commented:
I rebased the PR onto devel, which resolved the outstanding issues.

In addition, the pre-existing issue of `semCustomPragma` crashing when `semOverloadedCall` returns an error (noticed by @saem) is now also fixed.

saem reviewed (approved):
Bors r+

zerbina commented:
bors r+
